### PR TITLE
[Feature] support SP for FLUX.2-dev

### PR DESCRIPTION
**Pull request by nuclearwu**
Signed-off-by: wuzhongjian <wuzhongjian_yewu@cmss.chinamobile.com>

PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Refer to https://github.com/vllm-project/vllm-omni/issues/1806, support SP (Ulysses & Ring) for FLUX.2-dev

## Test Plan
Text-to-Image:
```
python3 examples/offline_inference/text_to_image/text_to_image.py \
  --model /workspace/mnt/cmss-johnjan/models/black-forest-labs/FLUX___2-dev/ \
  --prompt "a lovely bunny holding a sign that says 'vllm-omni'" \
  --seed 42 \
  --tensor-parallel-size 1 \
  --num-images-per-prompt 1 \
  --num-inference-steps 50 \
  --guidance-scale 4.0 \
  --height 1024 \
  --width 1024 \
  --enable-cpu-offload \
  --ulysses-degree 2 \
  --ring-degree 2 \
  --output outputs/flux2-dev-sp.png
```

Image-to-Image:
```
python examples/offline_inference/image_to_image/image_edit.py \
    --model /workspace/mnt/cmss-johnjan/models/black-forest-labs/FLUX___2-dev/ \
    --image outputs/flux2-dev.png \
    --prompt "replace the bunny in the image with dog." \
    --output outputs/flux2-dev-edit-sp.png \
    --seed 42 \
    --tensor-parallel-size 1 \
    --enable-cpu-offload \
    --ulysses-degree 2 \
    --ring-degree 2 \
    --num-inference-steps 50 \
    --guidance-scale 4.0
```

## Test Result
vLLM-Omni:
Reproduced with 4xA800.
Text-to-Image:
| Configuration | Ulysses degree | Ring degree | Generation Time | Speedup | Images Generated |
| :-: | :-: | :-: | :-: | :-: | :-: |
| Baseline | 1 | 1 | 89.0845s | 1.00x | flux2-dev |
| Ulysses | 4 | 1 | 71.9688s | 1.24x | flux2-dev |
| Ring | 1 | 4 | 70.8004s | 1.26x | flux2-dev |
| Hybrid Ulysses + Ring | 2 | 2 | 68.0736s | 1.31x | flux2-dev |

Image-to-Image:
* Target image:
flux2-dev

| Configuration | Ulysses degree | Ring degree | Generation Time | Speedup | Images Generated |
| :-: | :-: | :-: | :-: | :-: | :-: |
| Baseline | 1 | 1 | 147.6604s | 1.00x | flux2-dev-edit |
| Ulysses | 4 | 1 | 81.7383s | 1.81x | flux2-dev-edit |
| Ring | 1 | 4 | 87.3565s | 1.69x | flux2-dev-edit |
| Hybrid Ulysses + Ring | 2 | 2 | 85.9638s | 1.72x | flux2-dev-edit |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by nuclearwu** [comment]
cc @hsliuustc0106 @wtomin @lishunyang12

File: vllm_omni/diffusion/models/flux2/flux2_transformer.py [hunk 1]
```diff
@@ -183,34 +191,82 @@ def forward(
...
-            sin = sin.to(query.dtype)
-            query = self.rope(query, cos, sin)
-            key = self.rope(key, cos, sin)
+            sp_size = self.parallel_config.sequence_parallel_size
+            use_sp_joint_attention = (
+                sp_size is not None
+                and sp_size > 1
+                and is_forward_context_available()
+                and not get_forward_context().split_text_embed_in_sp
+            )
```
Comment (alex-jw-brooks) [inline vllm_omni/diffusion/models/flux2/flux2_transformer.py:223]: @nuclearwu are there any changes that don't match flux2klein's implementation?

https://github.com/vllm-project/vllm-omni/pull/2188 will combine two since they share the same DiT, so unifying them will also align acceleration support (including SP support for flux2dev)

Comment (nuclearwu) [reply vllm_omni/diffusion/models/flux2/flux2_transformer.py:223]: @alex-jw-brooks OK, that's great, the difference is that `Flux2Transformer2DModel` forward needs to take into account `is_torch_npu_available` for Flux.2-dev.

**Review by hsliuustc0106 — COMMENTED** [review]
## Review: [Feature] support SP for FLUX.2-dev (#3244)

### General

3 files, 247 additions, 45 deletions. Small, focused PR. `mergeStateStatus: BLOCKED` (all visible checks pass — likely an infra gate issue, same as other recent PRs).

---

### Test Evidence

This is a model card of how to present test results in a PR:
- 4 configurations tested (Baseline, Ulysses=4, Ring=4, Hybrid 2+2)
- Both T2I and I2I workflows
- Hardware stated (4×A800)
- Visible image outputs for every config
- Quantified speedups: **up to 1.31× (T2I) / 1.81× (I2I)**

The PR description content is strong. (The checklist template at the bottom is unchecked, but the actual content is there.)

---

### Code Assessment

The implementation follows the established SP pattern used by other diffusion models in the codebase:

- `_sp_plan` class variable defining what gets sharded/replicated
- `Flux2RopePrepare` cleanly separates RoPE computation from the forward pass so the SP plan can handle text (replicated) vs image (sharded) components independently
- `Flux2Attention` and `Flux2ParallelSelfAttention` gain an SP-aware joint-attention path using `AttentionMetadata(joint_query=..., joint_key=..., joint_value=..., joint_strategy="front")`
- Mask creation for SP-padded positions is correct

The `od_config` parameter flows cleanly through the constructor chain without breaking backward compatibility (defaults to `None` → `DiffusionParallelConfig()`).

---

### Issues

**1. `encoder_hidden_states_mask` is always `None`**

In `Flux2Transformer2DModel.forward()`:
```python
encoder_hidden_states_mask = None
```
and it's never set to a real value. If text embeddings can also be part of an SP-padded sequence, this mask should be created alongside `hidden_states_mask`. Is this intentionally skipped because text is always replicated (never sharded)?

**2. `hidden_states_mask.all()` guard masks an edge case**

```python
if hidden_states_mask.all():
    hidden_states_mask = None
```

If padding accidentally covers the entire sequence (shouldn't happen, but is a degenerate input), the mask is silently dropped and no masking occurs. Consider using a more explicit guard, e.g. `if not hidden_states_mask.any():` or logging a warning when the mask is all-true.

**3. NPU path uses `.cpu()` with `rope_prepare`**

The NPU path passes `.cpu()` tensors to `rope_prepare` and then moves each output back to NPU individually. The non-NPU path doesn't do `.cpu()`. Is this intentional? The previous code also did `.cpu()` for the NPU path, so it's a pre-existing pattern, not a regression — just flagging it.

---

### Non-blocking

**4. `_sp_plan` indices tied to `Flux2RopePrepare` output order**

The docstring on `Flux2RopePrepare` notes this correctly. If the output order changes, the `_sp_plan` breaks silently. Consider adding a guard or test that asserts `_sp_plan` indices match the `Flux2RopePrepare.forward` return signature.

**5. `apply_rope_to_qk` helper used in non-SP path but not SP path**

The non-SP paths use the helper, but the SP paths call `self.rope` directly. Minor inconsistency — not blocking since the SP path needs to split text/image RoPE separately anyway.

---

### Verdict

Clean implementation, excellent test evidence, well-structured. The main code concerns are the `encoder_hidden_states_mask` always being `None` (intentional?) and the `hidden_states_mask.all()` guard. Please address the merge gate issue.

**Comment by hsliuustc0106** [comment]
resolve conflicts please

**Comment by nuclearwu** [comment]
> resolve conflicts please

@hsliuustc0106 Done, please review again

**Comment by nuclearwu** [comment]
cc @hsliuustc0106

**Comment by hsliuustc0106** [comment]
generally it looks good to me, could you help update or add the recipe in a follow-up PR with the recommended cli configs

File: vllm_omni/diffusion/models/flux2/flux2_transformer.py [hunk 2]
```diff
@@ -722,6 +894,9 @@ def forward(
         joint_attention_kwargs = joint_attention_kwargs or {}
 
         num_txt_tokens = encoder_hidden_states.shape[1]
+        sp_size = self.parallel_config.sequence_parallel_size
+        if sp_size is not None and sp_size > 1:
+            get_forward_context().split_text_embed_in_sp = False
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/flux2/flux2_transformer.py:899]: Guard `get_forward_context()` here the same way you do in the attention modules and mask path. Without `is_forward_context_available()`, a direct `Flux2Transformer2DModel` call with `sequence_parallel_size > 1` outside the worker forward context can assert before inference starts. Klein has the same pattern, so not a regression — but easy to harden while you're here.

```suggestion
        if sp_size and sp_size > 1 and is_forward_context_available():
            get_forward_context().split_text_embed_in_sp = False
```

File: vllm_omni/diffusion/models/flux2/flux2_transformer.py [hunk 3]
```diff
@@ -543,6 +652,47 @@ def forward(self, ids: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
         return freqs_cos, freqs_sin
 
 
+class Flux2RopePrepare(nn.Module):
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/flux2/flux2_transformer.py:655]: Nit: docstring says "Flux.2-klein" — should be "Flux.2-dev" (or generic "Flux.2") since this is the dev transformer file.

**Review by hsliuustc0106 — APPROVED** [review]
LGTM. SP implementation matches the FLUX.2-klein pattern (`_sp_plan`, `Flux2RopePrepare`, joint attention, padding masks). Docs updated. Manual T2I/I2I evidence on 4×A800 is solid.

One question on I2I: each SP config shows a different output image despite `seed=42` — expected (distributed numerics / edit sensitivity), or wrong attachment? T2I parity looks good (same image across configs).

Follow-up (non-blocking): recipe PR with recommended `--ulysses-degree` / `--ring-degree` CLI configs for FLUX.2-dev.

File: vllm_omni/diffusion/models/flux2/flux2_transformer.py (hunk 2 above)
Comment (nuclearwu) [reply vllm_omni/diffusion/models/flux2/flux2_transformer.py:899]: Done

File: vllm_omni/diffusion/models/flux2/flux2_transformer.py (hunk 3 above)
Comment (nuclearwu) [reply vllm_omni/diffusion/models/flux2/flux2_transformer.py:655]: Done

**Comment by nuclearwu** [comment]
> generally it looks good to me, could you help update or add the recipe in a follow-up PR with the recommended cli configs

@hsliuustc0106 Of course, I'm just about to do this.

**Comment by nuclearwu** [comment]
> LGTM. SP implementation matches the FLUX.2-klein pattern (`_sp_plan`, `Flux2RopePrepare`, joint attention, padding masks). Docs updated. Manual T2I/I2I evidence on 4×A800 is solid.
> 
> One question on I2I: each SP config shows a different output image despite `seed=42` — expected (distributed numerics / edit sensitivity), or wrong attachment? T2I parity looks good (same image across configs).
> 
> Follow-up (non-blocking): recipe PR with recommended `--ulysses-degree` / `--ring-degree` CLI configs for FLUX.2-dev.

@hsliuustc0106 I test each SP config that shows a same output image in the previous main branch (0.19.0). It might be a problem with the accuracy of the latest branch.

**Comment by RuixiangMa** [comment]
@nuclearwu Could you add e2e tests for the SP support? For example, in test_flux_2_dev_expansion.py, adding cases with --ulysses-degree 2 and --ring-degree 2

**Comment by nuclearwu** [comment]
> @nuclearwu Could you add e2e tests for the SP support? For example, in test_flux_2_dev_expansion.py, adding cases with --ulysses-degree 2 and --ring-degree 2

@RuixiangMa Done
